### PR TITLE
[booster] implemented mixed precision class

### LINKED ISSUE #3054: [booster] implement precision
# Overview

The `Precision` is abstraction for training (mixed) precision and manages relevant functions such as grad scaler and clipping.

# Goal

An sample class definition is given below and is subject to changes.

```python
class Precision:
    def __init__(self, precision_type: dtype, grad_clipping_type: str, grad_clipping_value: float):
        self.precision_type = precision_type
        self.grad_clipping_type = grad_clipping_type
        self.grad_clipping_value = grad_clipping_value
        
    def setup_model(self, model) -> nn.Module:
        pass
        
    def setup_optimizer(self, optimizer) -> Optimizer:
        # inject grad clipping and unscale loss
        pass
        
    def scale_loss(self, loss) -> torch.Tensor:
        pass
```

### PR DESCRIPTION
**Pull request by FrankLeeeee**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

Fixed #3054 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

This PR implemented the `MixedPrecision` components for the booster module. The API design is slightly different from that specified in #3054  with the following changes:
1. created interface classes
2. let optimizer handle gradient clipping
3. modified the setup API for future extension like apex

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/booster.py [hunk 1]
```diff
@@ -1,37 +1,93 @@
...
 
 class Booster:
+    """
+    Booster is a high-level API for training neural networks. It provides a unified interface for
+    training with different precisio, accelerator, and plugin.
+
+    Examples:
+        >>> colossalai.launch(...)
+        >>> plugin = GeminiPlugin(stage=3, ...)
+        >>> engine = Booster(precision='fp16', plugin=plugin)
```
Comment (ver217) [inline colossalai/booster/booster.py:25]: should we replace engine with booster?

File: colossalai/booster/booster.py [hunk 2]
```diff
@@ -1,37 +1,93 @@
...
+            raise ValueError(
+                f'Expected the argument mixed_precision to be a string or an instance of Precision, but got {type(mixed_precision)}.'
+            )
 
-    def boost(
-        self, *args: Union[nn.Module, Optimizer, LRScheduler, DataLoader]
-    ) -> List[Union[nn.Module, Optimizer, LRScheduler, DataLoader]]:
-        # TODO: implement this method
-        pass
+    def boost(self, model: nn.Module, optimizer: Optimizer, criterion: Callable, lr_scheduler: LRScheduler,
```
Comment (ver217) [inline colossalai/booster/booster.py:68]: Can this handle multi-model using one optimizer? Or pure inference?

File: colossalai/booster/booster.py (hunk 1 above)
Comment (FrankLeeeee) [reply colossalai/booster/booster.py:25]: Done.

File: colossalai/booster/mixed_precision/fp16_torch.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    def __init__(self, module: nn.Module):
+        super().__init__()
+        self.module = module
+
+    def forward(self, *args, **kwargs):
+        with torch.cuda.amp.autocast():
+            return self.module(*args, **kwargs)
+
+
+class FP16_Torch_MixedPrecision(MixedPrecision):
```
Comment (ver217) [inline colossalai/booster/mixed_precision/fp16_torch.py:87]: Why this class name uses both camel case and snake case?

Comment (FrankLeeeee) [reply colossalai/booster/mixed_precision/fp16_torch.py:87]: I was think about this as well, `FP16TorchMixedPrecision`? Would it be difficult to read?

Comment (FrankLeeeee) [reply colossalai/booster/mixed_precision/fp16_torch.py]: @ver217  I have changed to camel case as advised by GPT-4 🤣

**Review by ver217 — APPROVED** [review]
